include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1138,6 +1136,16 @@ struct __early_exit_find_or
                 // So we use bool variable state check in the for-loop header.
                 __something_was_found = true;
             }
+
+            // Share found into state between items in our sub-group to early exit if something was found
+            // - this approach is applicable only for __parallel_or_tag (hen we search for any matching data entry)
+            // - for __parallel_find_forward_tag and __parallel_find_backward_tag we should process all data
+            if constexpr (_OrTagType{})
```

Comment:
I believe we should be able to use this subgroup coordinated early exit for all cases, not just `_OrTagType`.  However, I think we are blocked currently by our access pattern.  Right now, each work-item iterates in the direction of the search, (front-to-back or back-to-front) so it is safe to early exit individual work-items. 

This synchronization makes each subgroup work in lock step, coordinating the results, and if it were loading contiguous blocks of the input data in the direction of the search, we could do this coordinated subgroup early exit safely as well for the other search types.  Right now, it is **not** loading elements in contiguous blocks of the subgroup size, but rather contiguous blocks of size 16, and striding each work-item by 16 for each iteration.   

For a subgroup of size 32, we would have a subgroup loading half its data from one contiguous section, and another half from a later contiguous section.  This means that we cannot use coordinated subgroup early-exit, because a different work item may find a "earlier" element on a different iteration.

The fix for this is to re-write the access pattern to load contiguous blocks based on subgroup size rather than hardcoded to 16.  Then we should be able to remove this constexpr branch and benefit from this optimization for all find algorithms.  This also may make loads faster because they should be contiguous blocks of data rather than gathering from two spots.

Reply:
I propose to fix this in the separate PR